rfcs/20191017-tfx-standardized-inputs.md

```diff
@@ -0,0 +1,689 @@
+proto.
+
+Similarly, an Arrow column can be translated to various TF feedables.
+[An extension to the pipeline schema](#tensorrepresentation) is proposed to for
+a user to express the intention for conversion.
+
+The conversion can be efficient (zero-copy) in certain cases. It is
+[discussed](#efficient-arrow-tensor-conversion) in the detailed design.
+
+## Standardized Inputs APIs
+
+We propose a set of APIs that TFX components will call, and need to be
+implemented for each of the supported combination of {physical, logical} format.
+
+```py
+class TFXIO(object):
+  """Abstract basic class of all Standardized TFX inputs API implementations."""
+  def __init__(
+      self, pipeline_env,
+      schema: Optional[tfmd.Schema]=None
+  ):
+    pass
+
+  @abc.abstractmethod
+  def BeamSource(self,
+                 projections: Optional[List[Text]]=None
+  ) -> beam.PTransform:
+    """Returns a beam PTransform that produces PCollection[pa.RecordBatch].
+
+    May NOT raise an error if the TFMD schema was not provided at construction time.
+
+    Args:
+      projections: if not None, only the specified subset of columns will be
+      read.
+    """
+
+  @abc.abstractmethod
+  def TensorAdapter(self) -> TensorAdapter:
+    """Returns a TensorAdapter that converts pa.RecordBatch to TF inputs.
+
+    May raise an error if the TFMD schema was not provided at construction time.
+    """
+
+  @abc.abstractmethod
+  def ArrowSchema(self) -> pyarrow.Schema:
+    """Returns the schema of the Arrow RecordBatch generated by BeamSource().
+
+    May raise an error if the TFMD schema was not provided at construction time.
+    """
+
+  @abc.abstractmethod
+  def TFDataset(self, ...) -> tf.data.Dataset:
+    """Returns a Dataset of TF inputs.
+
+    May raise an error if the TFMD schema was not provided at construction time.
+    """
+```
+
+Where `TensorAdapter` is:
+
+```py
+class TensorAdapter(object):
+
+  def __init__(
+      self,
+      tensor_representations: Dict[str, TensorRepresentation]):
+    """Initializer.
+
+    Args:
+      tensor_representations: keys are the names of the output tensors; values
+      describe how an output tensor should be derived from a RecordBatch.
+    """
+    pass
+
+  def TypeSpecs(self) -> Dict[str, tf.TypeSpec]:
+    """Returns tf.TypeSpec for each tensor to be produced by ToBatchTensors().
+
+    TypeSpecs can be used to construct placeholders or tf.function signatures.
+    """
+
+  def ToBatchTensors(
+      self, record_batch: pyarrow.RecordBatch,
+      projections: Optional[List[TensorName]]=None
+  ) -> Dict[str, TFFeedable]:  # TFFeedable: np.ndarrays or tf.EagerTensor
+                               # (or compositions of them, i.e.
+                               # CompositeTensors).
+    """Converts a record batch to batched tensors.
+
+    Each will conform to the corresponding TypeSpec.
+
+    Args:
+      projections: if not None, only specified subset of tensors will be
+      converted.
```

Comment:
Can you explain this? An example would help. How are subsets specified?

Reply:
Revised the doc. Did it help?